File: 0037-sudoku-solver/0037-sudoku-solver.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solveSudoku = function(board) {
    function is_valid(board, row, col, num) {
        for (let i = 0; i < 9; i++) {
            const m = 3 * Math.floor(row / 3) + Math.floor(i / 3);
            const n = 3 * Math.floor(col / 3) + i % 3;
            if (board[row][i] == num || board[i][col] == num || board[m][n] == num) {
                return false; // not a valid number
            }
        }
        return true; // number is valid
    }

    function solve(board) {
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    for (let num = 1; num <= 9; num++) {
                        if (is_valid(board, i, j, String(num))) {
                            board[i][j] = String(num);
                            if (solve(board)) {
                                return true; // solve the rest of the board
                            }
                            board[i][j] = '.'; // undo the current cell for backtracking
                        }
                    }
                    return false; // no valid number can be placed at this cell
                }
            }
        }
        return true; // entire board is solved
    }

    solve(board); // invoke the solve function
};

File: 0037-sudoku-solver/0037-sudoku-solver.c
bool is_valid(char** board, int row, int col, char c) {
    for (int i = 0; i < 9; i++) {
        if (board[i][col] == c) return false; // check column
        if (board[row][i] == c) return false; // check row
        if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c) return false; // check 3x3 box
    }
    return true;
}

bool solve(char** board) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            if (board[i][j] == '.') {
                for (char c = '1'; c <= '9'; c++) {
                    if (is_valid(board, i, j, c)) {
                        board[i][j] = c; // place c for this cell

                        if (solve(board))
                            return true; // sudoku solved
                        else
                            board[i][j] = '.'; // undo the choice for this cell
                    }
                }
                return false; // no solution for this cell, backtrack
            }
        }
    }
    return true; // sudoku solved
}

void solveSudoku(char** board, int boardSize, int* boardColSize){
    solve(board);
}
